src/api/weth/WethApi.ts

```diff
@@ -0,0 +1,116 @@
+import Web3 from 'web3'
+
+import { Network, WithTxOptionalParams, Receipt } from 'types'
+import { WETH_ADDRESS_MAINNET, WETH_ADDRESS_RINKEBY } from 'const'
+import { wethAbi } from '@gnosis.pm/dex-js'
+import { logDebug } from 'utils'
+
+import { WethContract } from '@gnosis.pm/dex-js/build-esm/contracts/WethContract'
+
+export interface WrapUnwrapParams extends WithTxOptionalParams {
+  networkId: number
+  userAddress: string
+  amount: string
+}
+
+/**
+ * Interfaces for Wrapping ETH and unwrapping WETH token
+ */
+export interface WethApi {
+  deposit(params: WrapUnwrapParams): Promise<Receipt>
+  withdraw(params: WrapUnwrapParams): Promise<Receipt>
+}
+
+export interface WethApiDependencies {
+  web3: Web3
+  fetchGasPrice(): Promise<string | undefined>
+}
+
+function getWethAddressByNetwork(networkId: number): string {
+  switch (networkId) {
+    case Network.Mainnet:
+      return WETH_ADDRESS_MAINNET
+    case Network.Rinkeby:
+      return WETH_ADDRESS_RINKEBY
+    default:
+      throw new Error(`WethApi was not deployed to network ${networkId}`)
+  }
+}
+
+/**
+ * Basic implementation of WETH API
+ * Note that there's already another API for all ERC20, so only the deposit/withdraw is implemented in this API
+ */
+export class WethApiImpl implements WethApi {
+  protected _contractPrototype: WethContract
+  private web3: Web3
+  private fetchGasPrice: WethApiDependencies['fetchGasPrice']
+  protected static _contractsCache: { [network: number]: { [address: string]: WethContract } } = {}
```

Comment:
I would drop the address part. Not like we'll ever have more than one WETH contract deployed per network, right?

```suggestion
  protected static _contractsCache: { [network: number]: WethContract } = {}
```

Reply:
Totally right!

Reply:
7bbb9a302e7ac90bc9ac386fb1ce910db9eefc7d